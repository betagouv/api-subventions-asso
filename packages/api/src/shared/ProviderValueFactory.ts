import ProviderValueAdapter from "./adapters/ProviderValueAdapter";

export default class ProviderValueFactory {
    static buildProviderValueAdapter(provider: string, dataDate: Date) {
        return <T>(value: T) => ProviderValueAdapter.toProviderValue(value, provider, new Date(dataDate))
    }

    static buildProviderValuesAdapter(provider: string, dataDate: Date) {
        return <T>(value: T) => ProviderValueAdapter.toProviderValues(value, provider, new Date(dataDate))
    }
}