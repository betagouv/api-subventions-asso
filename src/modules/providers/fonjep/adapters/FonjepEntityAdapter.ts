import ProviderValueAdapter from "../../../../shared/adapters/ProviderValueAdapter";
import { siretToNIC } from "../../../../shared/helpers/SirenHelper";
import DemandeSubvention from "../../../demandes_subventions/interfaces/DemandeSubvention";
import Etablissement from "../../../etablissements/interfaces/Etablissement";
import FonjepRequestEntity from "../entities/FonjepRequestEntity";

export default class FonjepEntityAdapter {
    static PROVIDER_NAME = "Fonjep"
    
    static toDemandeSubvention(entity: FonjepRequestEntity): DemandeSubvention {
        const dataDate = entity.indexedInformations.date_versement
        return {
            service_instructeur: ProviderValueAdapter.toProviderValue(entity.indexedInformations.service_instructeur, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            status: ProviderValueAdapter.toProviderValue(entity.indexedInformations.status, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            annee_demande: ProviderValueAdapter.toProviderValue(entity.indexedInformations.annee_demande, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            financeur_principal: ProviderValueAdapter.toProviderValue(entity.indexedInformations.financeur_principal, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            montants: {
                accorde: ProviderValueAdapter.toProviderValue(entity.indexedInformations.montant_paye, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            }
        }
    }

    static toEtablissement(entity: FonjepRequestEntity): Etablissement {
        const dataDate = entity.indexedInformations.date_versement

        return {
            siret: ProviderValueAdapter.toProviderValues(entity.legalInformations.siret, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            nic: ProviderValueAdapter.toProviderValues(siretToNIC(entity.legalInformations.siret), FonjepEntityAdapter.PROVIDER_NAME, dataDate),
            adresse: ProviderValueAdapter.toProviderValues({
                code_postal: entity.indexedInformations.code_postal,
                commune: entity.indexedInformations.ville
            }, FonjepEntityAdapter.PROVIDER_NAME, dataDate),
        }
    }
}