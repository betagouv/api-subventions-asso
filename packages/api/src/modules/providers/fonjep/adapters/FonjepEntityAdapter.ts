import FonjepTiersEntity from "../entities/FonjepTiersEntity";
import FonjepPosteEntity from "../entities/FonjepPosteEntity";
import FonjepVersementEntity, { PayedFonjepVersementEntity } from "../entities/FonjepVersementEntity";
import FonjepTypePosteEntity from "../entities/FonjepTypePosteEntity";
import FonjepDispositifEntity from "../entities/FonjepDispositifEntity";
import FonjepDispositifDto from "../../../../dataProviders/db/providers/fonjep/dto/fonjepDispositifDto";
import { FonjepPosteDtoWithJSDate } from "../../../../dataProviders/db/providers/fonjep/dto/fonjepPosteDto";
import FonjepTypePosteDto from "../../../../dataProviders/db/providers/fonjep/dto/fonjepTypePosteDto";
import { FonjepVersementDto } from "../../../../dataProviders/db/providers/fonjep/dto/fonjepVersementDto";
import FonjepTiersDto from "../../../../dataProviders/db/providers/fonjep/dto/fonjepTiersDto";
import Siret from "../../../../identifierObjects/Siret";
import Ridet from "../../../../identifierObjects/Ridet";
import Rid from "../../../../identifierObjects/Rid";
import Siren from "../../../../identifierObjects/Siren";
import { GenericAdapter } from "../../../../shared/GenericAdapter";
import { FonjepApplicationFlatEntity, FonjepPaymentFlatEntity } from "../entities/FonjepFlatEntity";
import { DataBretagneRecords } from "../../dataBretagne/@types/DataBretagne";
import dataBretagneService from "../../dataBretagne/dataBretagne.service";
import { getShortISODate, modifyDateYear } from "../../../../shared/helpers/DateHelper";
import { removeWhitespace } from "../../../../shared/helpers/StringHelper";
import EstablishmentIdentifier from "../../../../identifierObjects/EstablishmentIdentifier";
import { ApplicationNature, ApplicationStatus, PaymentCondition } from "dto";

/**
 * Some of the nullIfEmpty calls have not been verified and were added base on every FonjepEntity type
 * Properties without nullIfEmpty are supposed to be mandatory / always present (checked in the 2024-12-31 extract file)
 */
export default class FonjepEntityAdapter {
    // duplicate with fonjepService.provider.id
    // see https://github.com/betagouv/api-subventions-asso/issues/3256
    static PROVIDER_NAME = "Fonjep";

    /**
     * This was defined a long time ago and must be updated and validated
     */
    private static FOUNDER_CODE_TO_BOP_MAPPER = {
        "10004": 163,
        "10005": 163,
        "10008": 147,
        "10009": 163,
        "10010": 209,
        "10012": 361,
        "10016": 163,
        "10017": 163,
    };

    /**
     *
     * @param code Financer code
     * It can be found in :
     * financeurCofinanceurCode (fonjepVersement)
     * financeurCode (fonjepDispositif)
     * financeurPrincipalCode (fonjepPoste)
     * code (fonjepTiers)
     *
     * @returns The state financial program number
     */
    static getBopFromFounderCode(code: number): number {
        return this.FOUNDER_CODE_TO_BOP_MAPPER[code];
    }

    static toFonjepTierEntity(tier: FonjepTiersDto, updateDate: Date): FonjepTiersEntity {
        return {
            code: tier["Code"],
            raisonSociale: tier["RaisonSociale"],
            estAssociation: tier["EstAssociation"],
            estCoFinanceurPostes: tier["EstCoFinanceurPostes"],
            estFinanceurPostes: tier["EstFinanceurPostes"],
            siretOuRidet: tier["SiretOuRidet"] ? removeWhitespace(tier["SiretOuRidet"]) : null,
            codePostal: tier["CodePostal"],
            ville: tier["Ville"],
            contactEmail: tier["ContactEmail"],
            updateDate,
        };
    }

    static toFonjepPosteEntity(poste: FonjepPosteDtoWithJSDate, updateDate: Date): FonjepPosteEntity {
        return {
            code: poste["Code"],
            annee: poste["Annee"],
            associationBeneficiaireCode: poste["AssociationBeneficiaireCode"],
            financeurPrincipalCode: poste["FinanceurPrincipalCode"],
            financeurAttributeurCode: poste["FinanceurAttributeurCode"],
            dateFinTriennalite: poste["DateFinTriennalite"],
            dispositifId: poste["DispositifId"],
            pstStatutPosteLibelle: poste["PstStatutPosteLibelle"],
            pstRaisonStatutLibelle: poste["PstRaisonStatutLibelle"],
            associationImplantationCode: poste["AssociationImplantationCode"],
            montantSubvention: poste["MontantSubvention"],
            pstTypePosteCode: poste["PstTypePosteCode"],
            pleinTemps: poste["PleinTemps"],
            doublementUniteCompte: poste["DoublementUniteCompte"],
            updateDate,
        };
    }

    static toFonjepVersementEntity(versement: FonjepVersementDto, updateDate: Date): FonjepVersementEntity {
        return {
            posteCode: versement["PosteCode"],
            periodeDebut: versement["PeriodeDebut"],
            periodeFin: versement["PeriodeFin"],
            dateVersement: versement["DateVersement"],
            montantAPayer: versement["MontantAPayer"],
            montantPaye: versement["MontantPaye"],
            updateDate,
        };
    }

    static toFonjepTypePosteEntity(typePoste: FonjepTypePosteDto, updateDate: Date): FonjepTypePosteEntity {
        return {
            code: typePoste["Code"],
            libelle: typePoste["Libelle"],
            updateDate,
        };
    }

    static toFonjepDispositifEntity(dispositif: FonjepDispositifDto, updateDate: Date): FonjepDispositifEntity {
        return {
            id: dispositif["ID"],
            libelle: dispositif["Libelle"],
            financeurCode: dispositif["FinanceurCode"],
            updateDate,
        };
    }

    public static extractPositionCode(entity: FonjepPaymentFlatEntity) {
        if (entity.ej !== GenericAdapter.NOT_APPLICABLE_VALUE) {
            throw new Error("You must extract a position code from a FonjepPaymentFlat entity");
        } else {
            // cf buildFlatPaymentId
            return entity.idVersement.split("-")[0];
        }
    }

    private static getConventionDate(position: FonjepPosteEntity): Date {
        if (position.dateFinTriennalite) return modifyDateYear(position.dateFinTriennalite, -3);
        throw new Error("We can't create FONJEP PaymentFlat without dateFinTriennalite");
    }

    private static buildFlatPaymentId(data: { thirdParty: FonjepTiersEntity; position: FonjepPosteEntity }) {
        const { thirdParty, position } = data;
        return `${position.code}-${getShortISODate(FonjepEntityAdapter.getConventionDate(position))}-${position.annee}-${
            thirdParty.siretOuRidet
        }`;
    }

    // this keeps the same structure as other providers payment flat uniqueId and add N/A for the missing fields
    private static buildPaymentFlatUniqueId(
        partialPaymentFlat: Omit<FonjepPaymentFlatEntity, "uniqueId">,
        beginningDate: Date,
    ) {
        const keys = [
            "fonjep",
            partialPaymentFlat.idVersement,
            partialPaymentFlat.programNumber,
            GenericAdapter.NOT_APPLICABLE_VALUE, // action code
            GenericAdapter.NOT_APPLICABLE_VALUE, // activity code
            getShortISODate(beginningDate), // was operationDate to be ISO with CHORUS but it can be identical for every payment of the same year. See positionCode S05543
            GenericAdapter.NOT_APPLICABLE_VALUE, // attachement comptable
            GenericAdapter.NOT_APPLICABLE_VALUE, // centre financier code
        ];
        return keys.join("-");
    }

    static toFonjepPaymentFlat(
        fonjepData: { payment: PayedFonjepVersementEntity; position: FonjepPosteEntity; thirdParty: FonjepTiersEntity },
        dataBretagneData: DataBretagneRecords,
    ): FonjepPaymentFlatEntity {
        const { payment, thirdParty, position } = fonjepData;
        if (!thirdParty.siretOuRidet) {
            throw new Error("Trying to create a FONJEP PaymentFlat without siret or ridet information");
        } else {
            const estabId = thirdParty.siretOuRidet;
            let estabIdType: "siret" | "ridet";
            let assoIdType: "siren" | "rid";
            let estabValueObjectId: Siret | Ridet;
            let assoValueObjectId: Siren | Rid;

            if (Siret.isSiret(estabId)) {
                estabIdType = "siret";
                assoIdType = "siren";
                estabValueObjectId = new Siret(estabId);
                assoValueObjectId = estabValueObjectId.toSiren();
            } else if (Ridet.isRidet(estabId)) {
                estabIdType = "ridet";
                assoIdType = "rid";
                estabValueObjectId = new Ridet(estabId);
                assoValueObjectId = estabValueObjectId.toRid();
            } else {
                throw new Error(
                    `Fonjep Tier with code ${position.code} has no siret or ridet and should not have been parsed`,
                );
            }

            const program =
                dataBretagneData.programs[
                    this.getBopFromFounderCode(Number(fonjepData.position.financeurPrincipalCode))
                ];

            const ministry = dataBretagneService.getMinistryEntity(program, dataBretagneData.ministries);

            const paymentId = this.buildFlatPaymentId({
                thirdParty,
                position,
            });

            const partialPaymentFlat = {
                idVersement: paymentId,
                exerciceBudgetaire: position.annee as number,
                typeIdEtablissementBeneficiaire: estabIdType,
                idEtablissementBeneficiaire: estabValueObjectId,
                typeIdEntrepriseBeneficiaire: assoIdType,
                idEntrepriseBeneficiaire: assoValueObjectId,
                amount: payment.montantPaye,
                operationDate: payment.dateVersement,
                centreFinancierCode: GenericAdapter.NOT_APPLICABLE_VALUE,
                centreFinancierLibelle: GenericAdapter.NOT_APPLICABLE_VALUE,
                attachementComptable: GenericAdapter.NOT_APPLICABLE_VALUE,
                regionAttachementComptable: GenericAdapter.NOT_APPLICABLE_VALUE,
                ej: GenericAdapter.NOT_APPLICABLE_VALUE,
                actionCode: GenericAdapter.NOT_APPLICABLE_VALUE,
                actionLabel: GenericAdapter.NOT_APPLICABLE_VALUE,
                activityCode: GenericAdapter.NOT_APPLICABLE_VALUE,
                activityLabel: GenericAdapter.NOT_APPLICABLE_VALUE,
                provider: this.PROVIDER_NAME.toLowerCase(),
                programName: program.label_programme,
                programNumber: program.code_programme,
                mission: program.mission,
                ministry: ministry?.nom_ministere || null,
                ministryAcronym: ministry?.sigle_ministere || null,
                updateDate: payment.updateDate,
            };

            // TODO: Another example where something nullable (Tier.FinanceurPrincipalCode) is required to build a unique ID
            // TODO: make FonjepTierEntity.financeurPrincipalCode mandatory and make it a number
            const uniqueId = this.buildPaymentFlatUniqueId(partialPaymentFlat, payment.periodeDebut);

            return { uniqueId, ...partialPaymentFlat };
        }
    }

    static toFonjepApplicationFlat(entities: {
        position: FonjepPosteEntity;
        beneficiary: FonjepTiersEntity; // association who made the application
        allocator: FonjepTiersEntity | undefined; // structure who handle the application payment
        instructor: FonjepTiersEntity | undefined; // structure who validates / instructs the application
        scheme: FonjepDispositifEntity | undefined;
    }): Omit<FonjepApplicationFlatEntity, "updateDate"> | null {
        const { position, beneficiary, allocator, instructor, scheme } = entities;
        if (!position.annee) throw new Error("FONJEP ApplicationFlat must have a budgetary year");
        if (!beneficiary.siretOuRidet) throw new Error("FONJEP ApplicationFlat must have a beneficiary siret or ridet");
        const beneficiaryEstablishmentIdType =
            EstablishmentIdentifier.getIdentifierType(beneficiary.siretOuRidet) || null;
        if (!beneficiaryEstablishmentIdType) {
            // TODO: this should not happen but has been encountered where rid was present in instead of ridet. See #3586
            console.log(
                `FONJEP ApplicationFlat must have a valid beneficiary siret or ridet, given ${beneficiary.siretOuRidet}.`,
            );
            return null;
        }

        const provider = this.PROVIDER_NAME.toLowerCase(); // replace this with #3338
        const applicationProviderId = `${position.code}-${getShortISODate(this.getConventionDate(position))}`;
        const applicationId = `${provider}-${applicationProviderId}`;
        const uniqueId = `${applicationId}-${position.annee}`;
        const paymentId = this.buildFlatPaymentId({ thirdParty: beneficiary, position });

        const allocatorIdType = allocator?.siretOuRidet
            ? EstablishmentIdentifier.getIdentifierType(allocator.siretOuRidet)
            : null;
        const instructiveDepartmentIdType = instructor?.siretOuRidet
            ? EstablishmentIdentifier.getIdentifierType(instructor.siretOuRidet)
            : null;

        return {
            uniqueId,
            applicationId,
            applicationProviderId,
            provider,
            joinKeyId: GenericAdapter.NOT_APPLICABLE_VALUE,
            joinKeyDesc: GenericAdapter.NOT_APPLICABLE_VALUE,
            allocatorName: allocator?.raisonSociale ?? null,
            allocatorId: allocator?.siretOuRidet ?? null,
            allocatorIdType,
            managingAuthorityName: "FONJEP",
            managingAuthorityId: null,
            managingAuthorityIdType: null,
            instructiveDepartmentName: instructor?.raisonSociale ?? null,
            instructiveDepartmentIdType,
            instructiveDepartementId: instructor?.siretOuRidet ?? null,
            beneficiaryEstablishmentId: beneficiary.siretOuRidet,
            beneficiaryEstablishmentIdType,
            budgetaryYear: position.annee,
            pluriannual: true,
            pluriannualYears: null, // null for now, see #3575 for updates
            decisionDate: null,
            conventionDate: this.getConventionDate(position),
            decisionReference: null,
            depositDate: null,
            requestYear: null,
            scheme: scheme?.libelle ?? null,
            subScheme: GenericAdapter.NOT_APPLICABLE_VALUE,
            statusLabel: ApplicationStatus.GRANTED, // always GRANTED because FONJEP only concerns payments (means that application has been granted)
            object: null,
            nature: ApplicationNature.MONEY,
            requestedAmount: position.montantSubvention,
            grantedAmount: position.montantSubvention,
            totalAmount: position.montantSubvention,
            ej: GenericAdapter.NOT_APPLICABLE_VALUE,
            paymentId,
            paymentCondition: PaymentCondition.PHASED,
            paymentConditionDesc: null,
            paymentPeriodDates: null,
            cofinancingRequested: GenericAdapter.NOT_APPLICABLE_VALUE,
            cofinancersNames: GenericAdapter.NOT_APPLICABLE_VALUE,
            cofinancersIdType: GenericAdapter.NOT_APPLICABLE_VALUE,
            confinancersId: GenericAdapter.NOT_APPLICABLE_VALUE,
            idRAE: null,
            ueNotification: null,
            subventionPercentage: null,
        };
    }
}
