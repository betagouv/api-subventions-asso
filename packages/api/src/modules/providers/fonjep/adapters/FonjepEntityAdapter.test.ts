import FonjepEntityAdapter from './FonjepEntityAdapter';
import FonjepEntity from "../../../../../tests/modules/providers/fonjep/__fixtures__/entity";
import ProviderValueFactory from '../../../../shared/ProviderValueFactory';


describe("FonjepEntityAdapter", () => {
    describe("toDemandeSubvention()", () => {
        const buildProviderValueAdapterMock = jest.spyOn(ProviderValueFactory, "buildProviderValueAdapter");
        it("should return a DemandeSubvention", () => {
            // @ts-expect-error: mock
            buildProviderValueAdapterMock.mockImplementationOnce(() => value => value)
            const actual = FonjepEntityAdapter.toDemandeSubvention(FonjepEntity);
            expect(actual).toMatchSnapshot();
        });
        it("should return a DemandeSubvension with co_financeur siret optionnal", () => {
            const entityWithInvalidCoFinanceurSiret = { ...FonjepEntity };
            entityWithInvalidCoFinanceurSiret.indexedInformations.co_financeur_siret = "";
            // @ts-expect-error: mock
            buildProviderValueAdapterMock.mockImplementationOnce(() => value => value)
            const demandeSubvention = FonjepEntityAdapter.toDemandeSubvention(entityWithInvalidCoFinanceurSiret);
            const expected = undefined;
            const actual = demandeSubvention.co_financement?.cofinanceur_siret;
            expect(actual).toEqual(expected);
        });
        it("should return a DemandeSubvention without co_financeur", () => {
            const entityWithoutCoFinanceur = { ...FonjepEntity };
            delete entityWithoutCoFinanceur.indexedInformations.co_financeur;
            // @ts-expect-error: mock
            buildProviderValueAdapterMock.mockImplementationOnce(() => value => value)
            const expected = undefined;
            const actual = FonjepEntityAdapter.toDemandeSubvention(entityWithoutCoFinanceur).co_financement;
            expect(actual).toEqual(expected);
        });
    });
    describe("toEtablissement()", () => {
        const buildProviderValuesAdapterMock = jest.spyOn(ProviderValueFactory, "buildProviderValuesAdapter");
        it("should return an Etablissement", () => {
            // @ts-expect-error: mock
            buildProviderValuesAdapterMock.mockImplementationOnce(() => value => [value]);
            const actual = FonjepEntityAdapter.toEtablissement(FonjepEntity);
            expect(actual).toMatchSnapshot();
        });
    });
});