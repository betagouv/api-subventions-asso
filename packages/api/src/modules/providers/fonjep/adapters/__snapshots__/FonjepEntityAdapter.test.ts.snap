// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FonjepEntityAdapter toCommon returns proper result 1`] = `
Object {
  "bop": "",
  "date_debut": 2022-02-02T00:00:00.000Z,
  "dispositif": "DISPOSITIF",
  "exercice": 2023,
  "montant_accorde": 43,
  "montant_verse": 42,
  "objet": "",
  "service_instructeur": "INSTRUCTEUR",
  "siret": "123456789",
  "statut": "Accordé",
}
`;

exports[`FonjepEntityAdapter toDemandeSubvention() should return a DemandeSubvention 1`] = `
Object {
  "annee_demande": 2022,
  "date_fin": 2022-01-03T00:00:00.000Z,
  "dispositif": "Dispositif",
  "montants": Object {
    "accorde": 500,
    "demande": 500,
  },
  "plein_temps": "Oui",
  "pluriannualite": "Oui",
  "service_instructeur": "XXX",
  "siret": "10000000000001",
  "status": "Attribué - Reconduction",
  "statut_label": "Accordé",
  "versementKey": "D00000",
}
`;

exports[`FonjepEntityAdapter toEtablissement() should return an Etablissement 1`] = `
Object {
  "adresse": Array [
    Object {
      "code_postal": "75000",
      "commune": "Paris",
    },
  ],
  "contacts": Array [
    Array [
      Object {
        "email": "contact@beta.gouv.fr",
      },
    ],
  ],
  "nic": Array [
    "00001",
  ],
  "siret": Array [
    "10000000000001",
  ],
}
`;

exports[`FonjepEntityAdapter toVersement() should return a Versement 1`] = `
Object {
  "amount": 1000,
  "bop": 163,
  "codePoste": "J00034",
  "dateOperation": 2022-01-02T00:00:00.000Z,
  "id": "unique_id",
  "montantAPayer": 1000,
  "periodeDebut": 2022-01-02T00:00:00.000Z,
  "periodeFin": 2022-01-02T00:00:00.000Z,
  "siret": "10000000000001",
  "versementKey": "J00034",
}
`;
